File: vscode/src/hough.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace std;

void cannyAndHough(cv::Mat src, cv::Mat dst, double rho, 
		double theta, int threshold, double minL, double minGap)
	{
	using namespace cv;

	Mat edges;
    Canny( src, edges, 50, 200, 3 );
    cvtColor( edges, dst, COLOR_GRAY2BGR );
    vector<Vec4i> lines;
    HoughLinesP( edges, lines, rho, theta, threshold, minL, minGap);
    for( size_t i = 0; i < lines.size(); i++ ) {
        line( dst, Point(lines[i][0], lines[i][1]),
        Point( lines[i][2], lines[i][3]), Scalar(0,0,255), 2, 8 );
    }
}

void readImage(char* name)
{
	cv::Mat image = cv::imread(name);
    int threshold = 80;
	double minL = 30;
	cv::Mat result = image.clone();

	if (image.empty())
	{
		cout << "画像ファイルを読み込めませんでした．" << endl;
		return;
	}

	cv::imshow("元画像", image);
    cannyAndHough(image, result, 1, CV_PI/180, threshold, minL, 10);
	cv::imshow("結果画像", result);

	cv::waitKey();
}

int main(int argc, char* argv[])
{
    if (argc < 2)
    {
        readImage("../image/building.png");
    }
	else{
        readImage(argv[1]);
    }
	
	return 0;
}

File: python/src/opencv/hough.py
import sys
import cv2
import numpy as np

class hough:

    def __init__(self,name):
        self.name = name
    
    
    def conversion(self):
        image = cv2.imread(self.name)
        cv2.imshow("元画像",image)
        gray = cv2.cvtColor(image,cv2.COLOR_BGR2GRAY)
        edges = cv2.Canny(gray,30,40,apertureSize=3)
        cv2.imshow("エッジ画像",edges)

        #長い線の検出
        lines = cv2.HoughLinesP(edges,1,np.pi/180,threshold=200,minLineLength=200,maxLineGap=50)
        #短い線の検出
        sLines = cv2.HoughLinesP(edges,1,np.pi/180,threshold=200,minLineLength=5,maxLineGap=5)

        dst = cv2.cvtColor(edges,cv2.COLOR_GRAY2BGR)
        for line in lines:
            x1,y1,x2,y2 = line[0]
            cv2.line(dst,(x1,y1),(x2,y2),(0,0,255),2)

        for line in sLines:
            x1,y1,x2,y2 = line[0]
            cv2.line(dst,(x1,y1),(x2,y2),(0,0,255),2)


        cv2.imshow("結果画像",dst)

        cv2.waitKey(0)
        cv2.destroyAllWindows()
    

args = sys.argv
if len(args) > 1:
    hough = hough(args[1])
else:
    hough = hough("./image/angle1.png")

hough.conversion()
